src/common-ssh/ssh.c

```diff
@@ -518,6 +520,64 @@ guac_common_ssh_session* guac_common_ssh_create_session(guac_client* client,
         return NULL;
     }
 
+    /* Check known_hosts, start by getting known_hosts file of user running guacd */
+    struct passwd *pw = getpwuid(getuid());
+    const char *known_hosts = strcat(pw->pw_dir, "/.ssh/known_hosts");
+    LIBSSH2_KNOWNHOSTS *ssh_known_hosts = libssh2_knownhost_init(session);
+    libssh2_knownhost_readfile(ssh_known_hosts, known_hosts, LIBSSH2_KNOWNHOST_FILE_OPENSSH);
+
```

Comment:
I agree that this is a departure from backward-compatibility, but I have two arguments for why I think it should be done that way, anyway:
- It makes Guacamole behave more like other SSH clients.  Most SSH clients require that you at least verify the host key on initial connection and agree to add it to the known_hosts file, and throw an error if it doesn't match.  This gets closer to that, though without prompting within the client, we can't ask if it doesn't initially exist.
- It makes SSH support in Guacamole behave more like RDP.  In RDP, you have to explicitly tell Guacamole to ignore server certificates, else it fails the connection unless you have correctly configured SSL on the system running guacd such that the certificates are trusted.  So, the behavior is more consistent within itself.

One potential compromise might be this: if the host key does not already exist in the known_hosts file, add it and allow the connection to continue instead of failing.  If the host key exists but does not match, fail.

Thoughts?  I definitely understand the importance of maintaining backward-compatibility; however, this seems like a case where perhaps breaking that would be acceptable, in some form or another.

Reply:
Okay, I've gone with the approach that known_hosts of the user running guacd will be read and checked against (in addition to any provided host key in the configuration parameters).  If there is a mismatch, the connection will fail.  If there is no match, guacd will warn but continue the connection.  Is that acceptable behavior?  Or would you still prefer an explicit enable parameter for checking the known_hosts file at all?

Reply:
> It makes Guacamole behave more like other SSH clients. Most SSH clients require that you at least verify the host key on initial connection and agree to add it to the known_hosts file, and throw an error if it doesn't match.

Closer, yes, but breaking SSH functionality for all existing users of Guacamole would be bad. The result of upgrading Guacamole from one version to another shouldn't be that all existing SSH connections (and VNC/RDP connections using SFTP) need to be reconfigured.

> It makes SSH support in Guacamole behave more like RDP.

I don't necessarily think that the built-in host checking of the FreeRDP library is the gold standard here, partly because in our case it universally requires that the keys be stored on the server running guacd. In fact, we should probably look toward implementing our own host checking within RDP, overriding FreeRDP's default behavior if a host key is specified, so that ignoring the certificate becomes unnecessary.

> One potential compromise might be this: if the host key does not already exist in the known_hosts file, add it and allow the connection to continue instead of failing.

I believe that would be a security vulnerability. The protections added through host checking would be able to be bypassed by a malicious user if they happen to use a guacd node which does not yet know about the SSH host in question.

> Thoughts? I definitely understand the importance of maintaining backward-compatibility; however, this seems like a case where perhaps breaking that would be acceptable, in some form or another.

My thoughts are:

* Existing VNC/RDP/SSH connections shouldn't break as a result of these changes
* If host checking is enabled (through providing a `known_hosts` file or through specifying the host key as a connection parameter), then strict host checking should be performed, with the connection being aborted if the host key is unknown or does not match.

Reply:
> Okay, I've gone with the approach that known_hosts of the user running guacd will be read and checked against (in addition to any provided host key in the configuration parameters).

Any reason behind checking against two sources? If a host key is specified via the connection parameters, shouldn't that override a local `known_hosts` entirely?

> Or would you still prefer an explicit enable parameter for checking the known_hosts file at all?

What about only checking local `known_hosts` if (1) the host key parameter is not provided and (2) the file exists?

Reply:
> Any reason behind checking against two sources? If a host key is specified via the connection parameters, shouldn't that override a local known_hosts entirely?
> What about only checking local known_hosts if (1) the host key parameter is not provided and (2) the file exists?

Sounds good to me...will make it so.

Reply:
Okay, I've implemented it as suggested and verified that it works.  The code seems a little clunky to me, but let me know what you think.  I was trying to avoid duplicating a lot of code checking it, but maybe the host key checking needs to be split into its own function?